cylc profile-battery

oliver-sanders opened the pull request:
Introduce `cylc profile-battery` command for comparing the overhead of running suites under different cylc versions. Close #1772 

`cylc profile-battery -e <experiment-ids> -v <version-ids>`

- Can only be run in a git repository.
  - Functional requirement for identifying cylc versions etc.
- Creates `cylc/.profile` directory (when run) for storing results etc.
- Experiments are json files which determine which suites to run, how to run them and how to profile them.
  - Currently these are json files, they could be upgraded to parsec if desired.
  - Built-in experiment files live in `cylc/dev/profile-experiments`, user experiment files live in `cylc/.profile/experiments`.
  - See `dev/profile-experiments/example` for info on experiment files.
  - Results are stored along with a checksum of the experiment file, once an experiment file is modified its results are orphaned (and can no-longer be plotted - this limitation could potentially be overcome) but can be reinstated using `cylc profile-battery promote <experiment-id>`.
  - Suites for profiling can be located anywhere accessible by the filesystem, built-in experiments should use suites in `dev/suites`.
- Version-ids can be any valid git identifier e.g. `6.11.2`, `HEAD`, `master`, `<some-commit>`.
- Profiling can be performed both on `run` and `validate`.
  - Profiling is performed using either `/usr/bin/time` and or `cylc <cmd> --profile`.
  - Only a few metrics are output to the user, to see more `cylc profile-battery -e <experiment_ids> -v <version_ids> --full`
- When run, profiling is performed then data is plotted (if matplotlib is installed) and output to `.profile/plots`.
- Results are saved to `.profile/results.json`, the experiment can then be run for other cylc versions and the results analysed.
  - Results file could be upgraded to results database if desired.
- To profile multiple cylc versions either check-out each one in turn then `cylc profile-battery -e <experiment> -v HEAD --no-plot` or specify all versions upfront and use the automatic `git checkout` feature `cylc profile-battery -e experiment -v HEAD 6.11.4 7.1.0`.

`cylc profile-battery --help` has some more details.

Other Changes:
- `dev/suites` are now compliant with cylc master (test added to ensure this).
- Profiling suites (including the "complex suite") added in `dev/suites`.

matthewrmshin commented on tests/profile-battery/00-compatability.t:
```diff
@@ -0,0 +1,74 @@
+else
+    echo -en "The following suites failed validation:\n${broken}" \
+        > "${TEST_NAME}.stderr"
+    cp "${TEST_NAME.stderr}" "${TEST_LOG_DIR}/${TEST_NAME}.stderr"
```
`"${TEST_NAME}.stderr"` instead of `"${TEST_NAME.stderr}"`.

matthewrmshin commented on tests/profile-battery/00-compatability.t:
```diff
@@ -0,0 +1,74 @@
+        broken="${suite}\n${broken}"
+    fi
+done
+if [[ ${broken} == "" ]]
```
`[[ -z "${broken}" ]]`

matthewrmshin commented on tests/profile-battery/00-compatability.t:
```diff
@@ -0,0 +1,74 @@
+then
+    ok "${TEST_NAME}"
+else
+    if [[ $? == 2 ]]
```
`elif`?

matthewrmshin commented on tests/profile-battery/00-compatability.t:
```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2016 NIWA
```
Check the copyright year, please.

matthewrmshin commented on lib/cylc/profiling/profile.py:
```diff
@@ -0,0 +1,313 @@
+                # Experiment failed to run, move onto the next one.
+                print 'Experiment "%s" failed at version "%s"' % (
+                    experiment['name'], version_id)
+                print exc
```
Error should go to STDERR.

matthewrmshin commented on lib/cylc/profiling/profile.py:
```diff
@@ -0,0 +1,313 @@
+                exp.get('mode', 'live'))
+            # Handle errors.
+            if err_file:
+                print 'WARNING: non-empty suite error log: ' + err_file
```
Error and warning messages should go to STDERR.

matthewrmshin commented on lib/cylc/profiling/git.py:
```diff
@@ -0,0 +1,88 @@
+                         stdout=PIPE).communicate()[0].strip()
+        if ancestor == commit1:
+            return True
+        return False
```
Can just say `return (ancestor == commit1)`.

matthewrmshin commented on lib/cylc/profiling/git.py:
```diff
@@ -0,0 +1,88 @@
+        cmd = ['git', 'checkout', branch]
+        print '$ ' + ' '.join(cmd)
+        check_call(cmd,
+                   stdout=open(os.devnull, 'wb'),
```
Can use `-q` for `git checkout`?

oliver-sanders replied:
Nice, didn't know about that one.

matthewrmshin commented on lib/cylc/profiling/git.py:
```diff
@@ -0,0 +1,88 @@
+        print '$ ' + ' '.join(cmd)
+        check_call(cmd,
+                   stdout=open(os.devnull, 'wb'),
+                   stderr=open(os.devnull, 'wb'))
```
Probably best to allow error message to print on failure.

matthewrmshin commented on bin/cylc-profile-battery:
```diff
@@ -0,0 +1,925 @@
+    * Supply cylc profile-battery with a complete list of the versions you wish
+      to profile, it will then provide the option to checkout the required
+      versions automatically.
+    * Checkout each version manually running cylc profile-battery against only
```
Instead of checkout, I wonder if it is better to create a separate tree using `git archive` or similar - so that the clone can still be used while the test is ongoing.

oliver-sanders replied:
I had though about checking archiving the requested versions to some temporary directory then using `CYLC_HOME_ROOT=<tmpdir>` and `CYLC_VERSION=<version>`. I'm not too sure why I didn't. I'm happy to go back and re-evaluate if you like.

matthewrmshin replied:
Yes, I think in the presence of the site wrapper, this is the way to do it.

Otherwise, I think all you need to do is create a work tree in a temporary location, and then *prepend* its `bin/` directory to `PATH` in the right environment.

matthewrmshin commented on .gitignore:
```diff
@@ -37,3 +37,5 @@ conf/*/*.rc.processed
 passphrase
 
+# profiling
+.profile
```
Can we think of a better name, given the usual meaning of this file name?

oliver-sanders replied:
Thought you might say that, will rename to `.profiling`.

matthewrmshin commented on bin/cylc-profile-battery:
```diff
@@ -0,0 +1,925 @@
+#!/usr/bin/env python
+"""Orchestrates experiments to profile the performance of cylc at different
```
Add usual header?

matthewrmshin commented:
(We have a proper Travis CI test failure here. Please fix.)

oliver-sanders commented:
Done.

matthewrmshin commented:
Travis CI still dying.

oliver-sanders commented:
This is back with me until I fix the travis failure.

matthewrmshin commented on tests/profile-battery/00-compatability.t:
```diff
@@ -0,0 +1,71 @@
+TEST_NAME="${TEST_NAME_BASE}-run-test-experiment"
+cylc profile-battery -e 'test' -v 'HEAD' --test \
+    2>"${TEST_LOG_DIR}/${TEST_NAME}.stderr"
+if [[ $? == 0 ]]
```
You may want to do:

```sh
RET_CODE=0
cylc profile-battery -e 'test' -v 'HEAD' --test \
    2>"${TEST_LOG_DIR}/${TEST_NAME}.stderr" || RET_CODE=$?
```

(If we put `set -e` on, an unprotected command returning non-zero will trigger error-exit.)

matthewrmshin commented on tests/profile-battery/00-compatability.t:
```diff
@@ -0,0 +1,71 @@
+    ok "${TEST_NAME}"
+elif [[ $? == 2 ]]
+then
+    1>&2 echo "Test requires git repository."
```
A bit more usual to put redirection at the end.

```sh
echo "Test requires git repository." >&2
```

matthewrmshin commented on lib/cylc/profiling/profile.py:
```diff
@@ -0,0 +1,330 @@
+
+        if to_purge:
+            print ('ERROR: The following suite(s) run directories could not be'
+                   ' deleted:')
```
Error to STDERR?

matthewrmshin commented on lib/cylc/profiling/profile.py:
```diff
@@ -0,0 +1,330 @@
+                except:
+                    # Analysis failed, move onto the next experiment.
+                    traceback.print_exc()
+                    print ('Analysis failed on results from experiment "%s"'
```
Error to STDERR?

matthewrmshin commented on lib/cylc/profiling/git.py:
```diff
@@ -0,0 +1,100 @@
+
+def is_git_repo():
+    """Returns true if we are currently within a git repository."""
+    proc = Popen('[ -d .git ] || git rev-parse --git-dir > /dev/null 2>&1',
```
Why not just run `git rev-parse` (i.e. no need to test for presence of `.git/`) and let `git` works it out?

matthewrmshin commented on lib/cylc/profiling/__init__.py:
```diff
@@ -0,0 +1,119 @@
+MEMORY_LINE_REGEX = re.compile('PROFILE: Memory: ([\d]+) KiB: ([\w.]+): (.*)')
+LOOP_MEMORY_LINE_REGEX = re.compile('(?:loop #|end main loop \(total loops )'
+                                    '([\d]+)(?:: |\): )(.*)')
+SLEEP_FUNCTION_REGEX = re.compile('([\d.]+)[\s]+[\d.]+[\s]+\{time.sleep\}')
```
For the above regular expressions, you may want to put some comments on the patterns you are trying to match and/or capture.

matthewrmshin commented:
(Travis CI still broken. Failure seems legit.)

oliver-sanders commented:
Test failure turned out to be due to Travis-ci using a different version of `cProfile` which yielded slightly different output.

@matthewrmshin Have addressed all feedback bar the git checkout vs archive question.

matthewrmshin commented:
The `git checkout` versus `git archive` issue is probably no big deal, as we can always perform the test using a sparse clone.

hjoliver commented:
@oliver-sanders - I still need to play with this a bit more, but v. nice work
